### ISSUE TITLE
Hard provisioned user can not be looked up in oidc backend in some cases

### ISSUE THREAD

**Issue by thlehmann-ionos** [nextcloud/user_oidc#898]
TL;DR: when the provider was configured with `unique-uid=1` (default) an ID would be created that is not equal to the ID in Nextcloud's backend and derived from the `sub` claim.

This would later lead to code not finding the user in the user_oidc backend, i.e. to check whether password authentication is possible.

### Configure and provision

1. Configure OIDC in the instance config for hard provisioning
2. Configure the OIDC provider with `--unique-uid` explicitly or implicitly true (1)
3. Provision the user in the Nextcloud backend with the ID from the `sub` claim
4. Provision the user in the user_oidc backend with the ID from the `sub` claim
   Note: user_oidc will add an entry to its user table with a [different ID][git-useroidc-idsvc].

#### Details: how the user was provisioned in Nextcloud

```
id="<ID as returned by the OIDC provider in the 'sub' claim>"
# other variables omitted

curl \
    -d "userid=${id}" \
    -d "password=" \
    -d "email=${email}" \
    -d "displayName=${display_name}" \
    -d "groups[]=${group}" \
    -H "OCS-APIRequest: true" \
        "http://${AUTH}@localhost:8080/ocs/v1.php/cloud/users"

```

#### Details: how user_oidc was configured and provisioned

The addon is configured with:

```
'user_oidc' => [
  'auto_provision' => false,
  'soft_auto_provision' => false,
],
```

The provider was configured as:

```
./occ user_oidc:provider "CLIENT" \
    --clientid="CLIENT" \
    --clientsecret="XXXXXXXXXXXXX" \
    --discoveryuri="DISCOVERY" \
    --scope="SCOPES"
```

Note: the provider was **not** configured with `--unique-uid=0`.

The user was provisioned with:

```
provider="<provider ID as returned by occ user_oidc:provider>"
id="<ID from the OIDC provider>"
# other variables omitted

curl \
    -D - \
    -X POST \
    -d "providerId=${provider}" \
    -d "userId=${id}" \
    -d "email=${email}" \
    -d "displayName=${display_name}" \
        "http://${AUTH}@localhost:8080/index.php/apps/user_oidc/user"

```

#### Why was the user provisioned in Nextcloud's and user_oidc's backend?

The call `/apps/user_oidc/user` (`ApiController->createUser()`) will fail at [`getUserFolder()`][git-useroidc-apictrlr], which tries to get a user folder from the Nextcloud backend. 

Thus, the assumption was that a user has to be provisioned in the Nextcloud backend too.

### Login via OIDC

1. Log user in via the OIDC provider
2. In the browser console test run or evaluate `window.backendAllowsPasswordConfirmation`
3. **Observation:** the value is true 
4. Wait a couple of hours
5. Go to settings and try to create an app token
6. **Expected:** the token gets created and is displayed
7. **Actual:** a password confirmation dialog is displayed (see below)

### How this leads to a problem

Later, when the user had logged in, the [user manager][git-nc-usrmgr] will query the all known backends (including Nextcloud's user database).

However, it would not be able to find the user in the user_oidc table, because [it was generated with a unique ID][git-useroidc-idsvc].

Since the user was not found in the user_oidc backend, Nextcloud will default to its own backend.

Later code will use the "wrong" backend and may mistakenly determine the user could authenticate themself.

Example [`JsConfigHelper`][git-nc-jscfghelper].

![password-confirmation-displayed-to-user-logged-in-via-oidc_](https://github.com/user-attachments/assets/8245d03c-749f-40a9-8d89-cc30661c4281)

### Debugging session

Earlier in the request the user would be looked up by the OIDC provider's ID (see type of the user mapper `$this` and the value of `$uid`):

![oidc-user-mapper-searching-for-idm-user-id_](https://github.com/user-attachments/assets/d3978748-0a76-4edb-bc29-f5e58cd89917)

Later - different from what's shown below - code would use the default Nextcloud backend (`$backend`) and fail the `instanceof IPasswordConfirmationBackend` test:

![jsconfighelper-testing-whether-the-backend-can-confirm-the-password-not-ok_2](https://github.com/user-attachments/assets/5663f99e-adc3-4066-814b-277dc7ef9d94)

This is how it would turn out if the provider was configured with `uniqueUid=0`, see how `$backend` is correct now:

![jsconfighelper-testing-whether-the-backend-can-confirm-the-password_](https://github.com/user-attachments/assets/57adf55e-8317-4155-a91f-3fde8534d076)

## Expectation

1. the documentation on hard provisioning should be extended to explain how to provision users in the Nextcloud backend correctly
2. _granted the updated documentation does not explain why this would be a user/admin/developer error:_ possible bug fix of the user_oidc provisioning code depending on a user to be added in the Nextcloud backend 
3. _granted the updated documentation does not explain why this would be a user/admin/developer error:_ user_oidc should not add users with a different ID than was specified through the API (since, as explained above, this may lead to subsequent errors)

The [code path during user provisioning][git-useroidc-idsvc] should either not exist or there should be documentation how to provision users in Nextcloud correctly when using `uniqueUid=1`.

[git-useroidc-idsvc]: https://github.com/nextcloud/user_oidc/blob/v5.0.3/lib/Service/LocalIdService.php#L30
[git-nc-usrmgr]: https://github.com/nextcloud/server/blob/v29.0.3/lib/private/User/Manager.php#L180
[git-nc-jscfghelper]: https://github.com/nextcloud/server/blob/v29.0.3/lib/private/Template/JSConfigHelper.php#L89
[git-useroidc-apictrlr]: https://github.com/nextcloud/user_oidc/blob/v5.0.3/lib/Controller/ApiController.php#L90

**Comment by artonge** [comment]
Regarding expectations:

1. Fair, the documentation can be updated
2. & 3. Would lead to an error down the road. When a user is logged in, the unique ID will be generated, so even if we change the behavior of the provisioning endpoint to provision users with their "original" IDs, then they won't be mapped correctly during login.

**Comment by artonge** [comment]
- Users would still be asked for password with `unique_id=1`, but they cannot answer a correct password.

**Comment by thlehmann-ionos** [comment]
In my case `unique-uid=0` is acceptable since we only want to use one Identity Provider. No blocker for me.

However, in general there's still a problem if customers have to use `unique-id=1` while using multiple providers (which might not necessarily be the case, but is possible). In this case, the shown problem from the "How this leads to a problem" section of the original description still exists, IMO.

Thank you for the fix #958, I'll look into that soon.
